### PR TITLE
Ported CLTV from BIP65 to PIVX. Added checks for supermajority activation.

### PR DESCRIPTION
**Pull request by nihilian**
activation.

### CODE REVIEW

**Comment by Fuzzbawls** [comment]
Unit tests are failing, and I don't see any actual block version enforcement being done once the supermajority threshold has passed.

File: src/script/interpreter.cpp [hunk 1]
```diff
@@ -4,6 +4,7 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include "main.h"
```
Comment (presstab) [inline src/script/interpreter.cpp]: We shouldn't be adding in dependency to main.h to this level of the code, it would be more appropriate to pass the proper flags through the arguments in the function.

For example, the check should be done outside of this function and only pass in `SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY` if supermajority is active.

**Comment by presstab** [comment]
@Fuzzbawls looks like changing this: https://github.com/PIVX-Project/PIVX/blob/master/src/main.cpp#L4226 to check against version 4 should do the trick.

File: src/chainparams.cpp [hunk 2]
```diff
@@ -131,9 +131,9 @@ class CMainParams : public CChainParams
         bnProofOfWorkLimit = ~uint256(0) >> 20; // PIVX starting difficulty is 1 / 2^12
         nSubsidyHalvingInterval = 210000;
         nMaxReorganizationDepth = 100;
-        nEnforceBlockUpgradeMajority = 750;
-        nRejectBlockOutdatedMajority = 950;
-        nToCheckBlockUpgradeMajority = 1000;
+        nEnforceBlockUpgradeMajority = 1080;
+        nRejectBlockOutdatedMajority = 1368;
+        nToCheckBlockUpgradeMajority = 1440;
```
Comment (Fuzzbawls) [inline src/chainparams.cpp]: What is the rationale behind this change? I know that the percentages are unaffected, just looking for the reasoning to extend the block window.

File: src/main.cpp [hunk 3]
```diff
@@ -3045,6 +3051,7 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
 
             std::vector<CScriptCheck> vChecks;
             unsigned int flags = SCRIPT_VERIFY_P2SH | SCRIPT_VERIFY_DERSIG;
+
```
Comment (Fuzzbawls) [inline src/main.cpp]: Could (should) be doing a check against supermajority within this function to append to the `flags` variable the `SCRIPT_VERIFY_CHECKLOCKTIMEVERIFY` status.

File: src/chainparams.cpp (hunk 2 above)
Comment (presstab) [reply src/chainparams.cpp]: There are 1440 blocks per day. Numbers changed to be in line with a lookback period of at least one day.

**Comment by presstab** [comment]
@Fuzzbawls what would the reasoning be to *not* prevent 1,400+ additional calculations per block?

**Comment by presstab** [comment]
Tests are failing because the serialized transactions added in the bitcoin suite have empty outputs which is something that is checked against in our codebase. Will need to replace the example transactions.

File: src/chainparams.cpp (hunk 2 above)
Comment (Fuzzbawls) [reply src/chainparams.cpp]: Ok, that’s what I thought, but didn’t want to take that assumption

**Comment by Warrows** [comment]
Is it WIP or ready for review?

**Comment by Fuzzbawls** [comment]
@nihilian can we get an update on this?
